[LLD] MinGW driver __image_base__ alternate name wrongly defined on i386 if no `-m` option from gcc

in lld/MinGW/Driver.cpp, there is:
https://github.com/llvm/llvm-project/blob/eb70253fcbe57c7cb3c309c06b94b05f2eab314d/lld/MinGW/Driver.cpp#L524-L527

This does not seem to work in the default case:
```console
$ g++ hello.cc -fuse-ld=lld
warning: failed to compute relocation: IMAGE_REL_I386_REL32, Invalid data was encountered while parsing the file
ld.lld: error: undefined symbol: __image_base__
>>> referenced by /t/r/msys2-runtime-3.3/src/msys2-runtime/winsup/cygwin/lib/_cygwin_crt0_common.cc:188
>>>               libmsys-2.0.a(_cygwin_crt0_common.o):(.text)
collect2: error: ld returned 1 exit status
```

As a workaround I can do:
```console
$ g++ hello.cc -fuse-ld=lld -Wl,-m,i386pe

$ ./a.exe
Hello
```

@mstorsjo @mati865

```console
$ g++ hello.cc -fuse-ld=lld -v
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/i686-pc-msys/13.3.0/lto-wrapper.exe
Target: i686-pc-msys
Configured with: /t/r/gcc/src/gcc-13.3.0/configure --build=i686-pc-msys --prefix=/usr --libexecdir=/usr/lib --enable-bootstrap --enable-static --enable-shared --enable-shared-libgcc --enable-version-specific-runtime-libs --with-arch=pentium4 --with-tune=generic --disable-multilib --disable-sjlj-exceptions --enable-__cxa_atexit --with-dwarf2 --enable-languages=c,c++,lto --enable-graphite --enable-threads=posix --enable-libatomic --enable-libgomp --disable-libitm --enable-libquadmath --enable-libquadmath-support --disable-libssp --disable-win32-registry --disable-symvers --with-gnu-ld --with-gnu-as --disable-isl-version-check --enable-checking=release --without-libiconv-prefix --without-libintl-prefix --with-system-zlib --enable-linker-build-id --enable-libstdcxx-filesystem-ts
Thread model: posix
Supported LTO compression algorithms: zlib
gcc version 13.3.0 (GCC)
COLLECT_GCC_OPTIONS='-fuse-ld=lld' '-v' '-shared-libgcc' '-mtune=generic' '-march=pentium4' '-dumpdir' 'a-'
 /usr/lib/gcc/i686-pc-msys/13.3.0/cc1plus.exe -quiet -v -idirafter /usr/lib/gcc/i686-pc-msys/13.3.0/../../../../include/w32api -idirafter /usr/lib/gcc/i686-pc-msys/13.3.0/../../../../i686-pc-msys/lib/../../include/w32api hello.cc -quiet -dumpdir a- -dumpbase hello.cc -dumpbase-ext .cc -mtune=generic -march=pentium4 -version -fuse-ld=lld -o /tmp/ccZxisAe.s
GNU C++17 (GCC) version 13.3.0 (i686-pc-msys)
        compiled by GNU C version 13.3.0, GMP version 6.3.0, MPFR version 4.2.1, MPC version 1.3.1, isl version isl-0.27-GMP

warning: MPFR header version 4.2.1 differs from library version 4.2.2.
GGC heuristics: --param ggc-min-expand=100 --param ggc-min-heapsize=131072
ignoring nonexistent directory "/usr/local/include"
ignoring nonexistent directory "/usr/lib/gcc/i686-pc-msys/13.3.0/../../../../i686-pc-msys/include"
ignoring duplicate directory "/usr/lib/gcc/i686-pc-msys/13.3.0/../../../../i686-pc-msys/lib/../../include/w32api"
#include "..." search starts here:
#include <...> search starts here:
 /usr/lib/gcc/i686-pc-msys/13.3.0/include/c++
 /usr/lib/gcc/i686-pc-msys/13.3.0/include/c++/i686-pc-msys
 /usr/lib/gcc/i686-pc-msys/13.3.0/include/c++/backward
 /usr/lib/gcc/i686-pc-msys/13.3.0/include
 /usr/lib/gcc/i686-pc-msys/13.3.0/include-fixed
 /usr/include
 /usr/lib/gcc/i686-pc-msys/13.3.0/../../../../include/w32api
End of search list.
Compiler executable checksum: 91915367b841ef4c4b6dae8707f1ee53
COLLECT_GCC_OPTIONS='-fuse-ld=lld' '-v' '-shared-libgcc' '-mtune=generic' '-march=pentium4' '-dumpdir' 'a-'
 /usr/lib/gcc/i686-pc-msys/13.3.0/../../../../i686-pc-msys/bin/as.exe -v -o /tmp/cczrWzTj.o /tmp/ccZxisAe.s
GNU assembler version 2.44 (i686-pc-msys) using BFD version (GNU Binutils) 2.44
COMPILER_PATH=/usr/lib/gcc/i686-pc-msys/13.3.0/:/usr/lib/gcc/i686-pc-msys/13.3.0/:/usr/lib/gcc/i686-pc-msys/:/usr/lib/gcc/i686-pc-msys/13.3.0/:/usr/lib/gcc/i686-pc-msys/:/usr/lib/gcc/i686-pc-msys/13.3.0/../../../../i686-pc-msys/bin/
LIBRARY_PATH=/usr/lib/gcc/i686-pc-msys/13.3.0/:/usr/lib/gcc/i686-pc-msys/13.3.0/../../../../i686-pc-msys/lib/:/usr/lib/gcc/i686-pc-msys/13.3.0/../../../:/usr/lib/
COLLECT_GCC_OPTIONS='-fuse-ld=lld' '-v' '-shared-libgcc' '-mtune=generic' '-march=pentium4' '-dumpdir' 'a.'
 /usr/lib/gcc/i686-pc-msys/13.3.0/collect2.exe -plugin /usr/lib/gcc/i686-pc-msys/13.3.0/msys-lto_plugin.dll -plugin-opt=/usr/lib/gcc/i686-pc-msys/13.3.0/lto-wrapper.exe -plugin-opt=-fresolution=/tmp/ccnG5Is6.res -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lmsys-2.0 -plugin-opt=-pass-through=-ladvapi32 -plugin-opt=-pass-through=-lshell32 -plugin-opt=-pass-through=-luser32 -plugin-opt=-pass-through=-lkernel32 -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc --build-id --wrap _Znwj --wrap _Znaj --wrap _ZdlPv --wrap _ZdaPv --wrap _ZnwjRKSt9nothrow_t --wrap _ZnajRKSt9nothrow_t --wrap _ZdlPvRKSt9nothrow_t --wrap _ZdaPvRKSt9nothrow_t -Bdynamic --dll-search-prefix=msys- --large-address-aware --tsaware -fuse-ld=lld /usr/lib/gcc/i686-pc-msys/13.3.0/../../../crt0.o /usr/lib/gcc/i686-pc-msys/13.3.0/crtbegin.o -L/usr/lib/gcc/i686-pc-msys/13.3.0 -L/usr/lib/gcc/i686-pc-msys/13.3.0/../../../../i686-pc-msys/lib -L/usr/lib/gcc/i686-pc-msys/13.3.0/../../.. /tmp/cczrWzTj.o -lstdc++ -lgcc_s -lgcc -lmsys-2.0 -ladvapi32 -lshell32 -luser32 -lkernel32 -lgcc_s -lgcc /usr/lib/gcc/i686-pc-msys/13.3.0/../../../default-manifest.o /usr/lib/gcc/i686-pc-msys/13.3.0/crtend.o
warning: failed to compute relocation: IMAGE_REL_I386_REL32, Invalid data was encountered while parsing the file
ld.lld: error: undefined symbol: __image_base__
>>> referenced by /t/r/msys2-runtime-3.3/src/msys2-runtime/winsup/cygwin/lib/_cygwin_crt0_common.cc:188
>>>               libmsys-2.0.a(_cygwin_crt0_common.o):(.text)
collect2: error: ld returned 1 exit status
```

d'oh, I saw and wondered about the entry usage of `"i386pe"` comparison, turns out I need to fix it too:
```
ld.lld: error: <root>: undefined symbol: ___msys_dll_entry@12
collect2: error: ld returned 1 exit status
clang++: error: linker (via gcc) command failed with exit code 1 (use -v to see invocation)
```

--- a/lld/MinGW/Driver.cpp
+++ b/lld/MinGW/Driver.cpp
@@ -171,6 +171,14 @@ searchLibrary(StringRef name, ArrayRef<StringRef> searchPaths, bool bStatic) {
   return "";
 }
 
+static bool isI386Target(const opt::InputArgList &args,
+                         const Triple &defaultTarget) {
+  auto *a = args.getLastArg(OPT_m);
+  if (a)
+    return StringRef(a->getValue()) == "i386pe";
+  return defaultTarget.getArch() == Triple::x86;
+}
+
 namespace lld {
 namespace coff {
 bool link(ArrayRef<const char *> argsArr, llvm::raw_ostream &stdoutOS,
@@ -216,6 +224,8 @@ bool link(ArrayRef<const char *> argsArr, llvm::raw_ostream &stdoutOS,
     return false;
   }
 
+  Triple defaultTarget(Triple::normalize(sys::getDefaultTargetTriple()));
+
   std::vector<std::string> linkArgs;
   auto add = [&](const Twine &s) { linkArgs.push_back(s.str()); };
 
@@ -224,7 +234,7 @@ bool link(ArrayRef<const char *> argsArr, llvm::raw_ostream &stdoutOS,
 
   if (auto *a = args.getLastArg(OPT_entry)) {
     StringRef s = a->getValue();
-    if (args.getLastArgValue(OPT_m) == "i386pe" && s.starts_with("_"))
+    if (isI386Target(args, defaultTarget) && s.starts_with("_"))
       add("-entry:" + s.substr(1));
     else if (!s.empty())
       add("-entry:" + s);
@@ -521,7 +531,7 @@ bool link(ArrayRef<const char *> argsArr, llvm::raw_ostream &stdoutOS,
   for (auto *a : args.filtered(OPT_Xlink))
     add(a->getValue());
 
-  if (args.getLastArgValue(OPT_m) == "i386pe")
+  if (isI386Target(args, defaultTarget))
     add("-alternatename:__image_base__=___ImageBase");
   else
     add("-alternatename:__image_base__=__ImageBase");

